### DISCUSSION TITLE
Issues  with using provided dockerfile

### DISCUSSION THREAD

**Discussion by Sne4kers** [skhatiri/UAV-Testing-Competition#6] [Submission] [answered]
I have gotten a lot of issues since the update for the submission. 

Firstly, after changing the `testcase.py` to a submission format, our generator stopped executing due to importing a local agent.
```
Traceback (most recent call last):
  File "cli.py", line 9, in <module>
    from split_grid_generator import DelftSplitSectorGenerator
  File "/root/submission/split_grid_generator.py", line 3, in <module>
    from testcase import TestCase
  File "/root/submission/testcase.py", line 8, in <module>
    from aerialist.px4.local_agent import LocalAgent
  File "/root/submission/venv/lib/python3.8/site-packages/aerialist/px4/local_agent.py", line 4, in <module>
    from .simulator import Simulator
  File "/root/submission/venv/lib/python3.8/site-packages/aerialist/px4/simulator.py", line 16, in <module>
    class Simulator(object):
  File "/root/submission/venv/lib/python3.8/site-packages/aerialist/px4/simulator.py", line 17, in Simulator
    PX4_DIR = config("PX4_HOME")
  File "/root/submission/venv/lib/python3.8/site-packages/decouple.py", line 243, in __call__
    return self.config(*args, **kwargs)
  File "/root/submission/venv/lib/python3.8/site-packages/decouple.py", line 105, in __call__
    return self.get(*args, **kwargs)
  File "/root/submission/venv/lib/python3.8/site-packages/decouple.py", line 90, in get
    raise UndefinedValueError('{} not found. Declare it as envvar or define a default value.'.format(option))
decouple.UndefinedValueError: PX4_HOME not found. Declare it as envvar or define a default value.
```
It works well when a local agent import is commented out. 

Secondly, nothing happens if the provided docker image is used to run the generator together with a new test case. When the generator reaches `test.execute()` I simply get nothing, logs are stopped and not even the message from the execute method is output. This has already been the case for more than an hour. No timeout happening nor funny message. I have also tried passing the Docker unix socket together with switching the agent type to docker - the following error message is printed:
```
aerialist.px4.docker_agent - ERROR - /bin/sh: 1: docker: not found

root - INFO - Failed at running testcase
root - INFO - 'DockerAgent' object has no attribute 'container_id'
Traceback (most recent call last):
  File "/src/generator/split_grid_generator.py", line 114, in generate
    distances = self.run_testcase(tests_to_run[i])
  File "/src/generator/split_grid_generator.py", line 61, in run_testcase
    test.execute()
  File "/src/generator/testcase.py", line 25, in execute
    agent = DockerAgent(self.test)
  File "/src/aerialist/aerialist/px4/docker_agent.py", line 36, in __init__
    self.docker_config = self.import_config()
  File "/src/aerialist/aerialist/px4/docker_agent.py", line 109, in import_config
    self.import_file(self.config.drone.mission_file, "/io/")
  File "/src/aerialist/aerialist/px4/docker_agent.py", line 155, in import_file
    f"docker cp '{src}' {self.container_id}:'{dest}'",
AttributeError: 'DockerAgent' object has no attribute 'container_id'
```

Additionally, I have tried building an image with a random generator with a given docker file. I have started it with `docker run --detach -it uav-random-generator sh -c "python3 cli.py generate case_studies/mission1.yaml 120"` and it also results in radio silence. 

Any tips on what should I do? The code used to run perfectly without a docker container with an older version of `testcase.py`.

**Comment by skhatiri** [comment, marked as answer]
it is strange since others have made it work. 
Can you try running the exact code from the pipeline (not including your test generator) and see if you still have the problem?
Let me know how it goes.

**Reply by Sne4kers** [reply]
It doesn't work. Simply doesn't produce any output on all of my Ubuntu machines.

**Reply by skhatiri** [reply]
even the dockered version!?
I am not sure I can help much more without seeing and trying to run your code. 

Since the deadline is close, probably the best way is to send us your code in two versions: 1) running and tested with the previous version of the pipeline, and 2) upgraded but not working with the new version.
We will contact you to get help running your code if needed. Send us an email to sync with details.

**Reply by soizhiwen** [reply]
I have the same problem.

**Reply by Sne4kers** [reply]
That's the issue it doesn't work in Docker. I used to run my generator before just by opening a new shell (using `screen` on Linux) and it worked fine. When I copied the new `testcase.py` it stopped working due to the import of the LocalAgent. Before, I was also using the standard `testcase.py` just with a DockerAgent and it worked fine. 

When a new `testcase.py` was introduced and I could run it through `screen`, I tried running it using the given Dockerfile via docker but that didn't work either, even for a random generator.

**Reply by skhatiri** [reply]
Just to make sure we are on the same page, I double-checked the dockerfile with the following steps:

- being on the master branch of [UAV-Testing-Competition](https://github.com/skhatiri/UAV-Testing-Competition/) without any modification
- `cd snippets/`
- `docker build . -t generator`
- `docker run -it generator` -> opens the terminal inside the docker container
- `python3 cli.py generate case_studies/mission1.yaml 3` -> run random generator with a budget of 3 for testing
here is the output, works:
![image](https://github.com/skhatiri/UAV-Testing-Competition/assets/16767722/81d9a64c-039e-4852-8eab-824162d1fbaa)

To clarify: 
- The local agent will not work on your own machine, you will still use the dockerAgent there. 
`AGENT = config("AGENT", default=AgentConfig.DOCKER)`
- Inside the Docker container, Local Agent is used to run the simulations already there, not in a separate Docker container. 
`ENV AGENT local`
- The default settings take care of all this.

In the end, as I said, if you keep having this issue, you can comment out the problematic lines or submit with the previous version.

**Reply by Sne4kers** [reply]
Yes, these instructions lead to empty logs in the docker container. I will submit the repo as is, and with the new version of the `testcase.py`, but I will also provide the instructions that allowed me to run my generator in `screen`.

As far as I know, I am not the only one having these issues, and it's specifically something with the integration of the Aeralist, Docker, and the generators, as generators with Aeralist running through the docker agent are perfect on my machine.

**Reply by dgumenyuk** [reply]
I have this issue too with `decouple.UndefinedValueError: PX4_HOME not found. Declare it as envvar or define a default value.`.  It seems like the .env file is not loading properly... When I run outside the docker (host cli) it works, if I comment out "LocalAgent".

**Reply by skhatiri** [reply]
@dgumenyuk having checked your code, I realized the issue. 
You are installing aerialist thought requirements.txt [here](https://github.com/swat-lab-optimization/UAV-Testing-Competition/blob/c915e80da6fdcb2966e93a276105d4946a389776/snippets/requirements.txt#L2C1-L2C46), but aerialist is already installed an configured for the Docker image. so this is breaking the configurations. 

So, **aerialist should not be added to requirement.txt** since the Dockerfile is taking care of its installation and proper configuration. I already updated the template to make this clear with [this commit](https://github.com/skhatiri/UAV-Testing-Competition/commit/0535c1ac574349e1508dbcf46bb108cfcbba1d15)

@Sne4kers maybe you have the same issue?

**Reply by skhatiri** [reply]
@Sne4kers [this commit](https://github.com/skhatiri/UAV-Testing-Competition/commit/50e3e983ef98f1be4a550922fc23b68b35740a4b) should have addressed your import issue for local agent.

**Comment by dgumenyuk** [comment]
@skhatiri Thank you very much for the reply. By removing the "Aerialist" from the requirements and adopting the new testcase.py the issue was resolved. I can run my test generator inside the container.

**Comment by Sne4kers** [comment]
> @dgumenyuk having checked your code, I realized the issue. You are installing aerialist thought requirements.txt [here](https://github.com/swat-lab-optimization/UAV-Testing-Competition/blob/c915e80da6fdcb2966e93a276105d4946a389776/snippets/requirements.txt#L2C1-L2C46), but aerialist is already installed an configured for the Docker image. so this is breaking the configurations.
> 
> So, **aerialist should not be added to requirement.txt** since the Dockerfile is taking care of its installation and proper configuration. I already updated the template to make this clear with [this commit](https://github.com/skhatiri/UAV-Testing-Competition/commit/0535c1ac574349e1508dbcf46bb108cfcbba1d15)
> 
> @Sne4kers maybe you have the same issue?

I don't have the Aeralist in the `requirements.txt`. The update helped fix the issue when running the generator not in the Docker - in a separate shell, but when running in Docker the issue is still there. I have added you to the repository of the generator if you are curious about investigating it, but I don't know what the issue might be.

**Reply by skhatiri** [reply]
For the moment it's fine for you to submit your test generator as is. We will check and contact you If needed to make it work also using Docker. This is because in the end, we will evaluate all generators in a Docker container.

**Comment by skhatiri** [comment]
@Sne4kers we were able to build the docker image for your submission without any changes to the Dockerfile or your code. So probably it was some issues with your setup that caused the issue. 
I am marking this discussion as answered then.
